docs/feature/dualstack/Cilium-dualstack.md

```diff
@@ -0,0 +1,69 @@
+[`AssignAnyAvailableIPConfig`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20(service%20*HTTPRestService)%20AssignAnyAvailableIPConfig(podInfo%20cns.PodInfo)%20(cns.PodIpInfo%2C%20error)%20%7B). This function then searches for the first available IP and returns it back up to the handler. The handler then packages the IP into a [`IPConfigResponse`](https://github.com/Azure/azure-container-networking/blob/bd299fe7271a7a23b3d0268d8e14ad812181e076/cns/NetworkContainerContract.go#L419) which gets returned to the client.
+## New Implementation
+
+### New Contract Type
+
+To preserve the current contract type but also to allow for the new functionality we will add a new type, `IPConfigsResponse` to the contract. This new type will be similar to the prexisting [`IPConfigResponse`](https://github.com/Azure/azure-container-networking/blob/bd299fe7271a7a23b3d0268d8e14ad812181e076/cns/NetworkContainerContract.go#L419) but replacing the single `PodIpInfo` with a slice. This will allow for a single call to the API to return one IP per NC.
+
+```diff
+// IPConfigResponse is used in CNS IPAM mode as a response to CNI ADD
+type IPConfigResponse struct {
+	PodIpInfo PodIpInfo
+	Response  Response
+}
+
++// IPConfigsResponse is used in CNS IPAM mode to return a slice of IP configs as a response to CNI ADD
++type IPConfigsResponse struct {
++	PodIpInfo []PodIpInfo
++	Response  Response
```

Comment:
I don't know what we use this `Response` type for, but if it exists in each `PodIpInfo`, do we also need one for the whole list?

Reply:
From my understanding I believe the response just is a return code and a message for the `IPConfigsResponse` and isn't specific to each `PodIpInfo`. If for some reason we get an error trying to get IPs it is adds a `ReturnCode` and the error message to `Response`

Reply:
@rjdenny you need to consider what happens if we get one IPConfig and then fail the next one. Is the first one marked as "Assigned" already? We would need to roll that back, and make the whole op transactional - entirely successful or entirely failed.

Reply:
@rbtr yes we had thought about it and have added a todo in the code as a reminder to address it before merging the change. We cannot have the partial success because in that case, we may mark IPs as 'Assigned' w/o the containers using those IPs. We cannot rely on the client to do the right thing of releasing the partially succeeded requests. We may run out of the IPs if we don't handle this ourselves.